src/basic/virt.c

```diff
@@ -796,6 +797,22 @@ Virtualization detect_container(void) {
         if (r < 0)
                 log_debug_errno(r, "Failed to detect cgroup namespace: %m");
 
+        /* Finally, the root pid namespace has an hardcoded inode number of 0xEFFFFFFC since kernel 3.8, so
+         * if all else fails we can check the inode number of our pid namespace and compare it. */
+        pidns_fd = namespace_open_by_type(NAMESPACE_PID);
+        if (pidns_fd < 0)
+                log_debug_errno(pidns_fd, "Failed to open PID namespace, ignoring: %m");
+        else {
+                struct stat st;
+
+                if (fstat(pidns_fd, &st) < 0)
+                        log_debug_errno(errno, "Failed to fstat pid namespace fd, ignoring: %m");
+                else if ((uint64_t) st.st_ino != 0xEFFFFFFC) {
```

Comment:
UINT64_C() ?

Reply:
yes please

Reply:
also, could given there's a kernel internal identifier for this, can we replicate this in userspace, maybe in missing_xyz.h? i.e.

```c
/* Root namespace inode numbers, as per include/linux/proc_ns.h in the kernel source tree*/
#define PROC_PID_INIT_INO UINT64_C(0xEFFFFFFC)
*/